beacon-chain/p2p/fork.go

```diff
@@ -0,0 +1,127 @@
+)
+
+// ENR key used for eth2-related fork data.
+const eth2EnrKey = "eth2"
+
+// EnrForkID represents a special value ssz-encoded into
+// the local node's ENR for discovery purposes. Peers should
+// only connect if their enrForkID matches.
+type EnrForkID struct {
+	CurrentForkDigest [4]byte
+	NextForkVersion   [4]byte
+	NextForkEpoch     uint64
+}
+
+// Compares fork ENRs between an incoming peer's record and our node's
+// local record values for current and next fork version/epoch.
+func (s *Service) compareForkENR(record *enr.Record) error {
+	currentRecord := s.dv5Listener.LocalNode().Node().Record()
+	peerForkENR, err := retrieveForkEntry(record)
+	if err != nil {
+		return err
+	}
+	currentForkENR, err := retrieveForkEntry(currentRecord)
+	if err != nil {
+		return err
+	}
+	// Clients SHOULD connect to peers with current_fork_digest, next_fork_version,
+	// and next_fork_epoch that match local values.
+	if peerForkENR.CurrentForkDigest != currentForkENR.CurrentForkDigest {
+		return fmt.Errorf(
+			"fork digest of peer: %v, does not match local value: %v",
+			peerForkENR.CurrentForkDigest,
+			currentForkENR.CurrentForkDigest,
+		)
+	}
+	// Clients MAY connect to peers with the same current_fork_version but a
+	// different next_fork_version/next_fork_epoch. Unless ENRForkID is manually
+	// updated to matching prior to the earlier next_fork_epoch of the two clients,
+	// these type of connecting clients will be unable to successfully interact
+	// starting at the earlier next_fork_epoch.
+	if peerForkENR.NextForkEpoch != currentForkENR.NextForkEpoch {
+		log.WithFields(logrus.Fields{
+			"peerNextForkEpoch": peerForkENR.NextForkEpoch,
+			"nodeNextForkEpoch": currentForkENR.NextForkEpoch,
+		}).Debug("Peer matches fork digest but has different next fork epoch")
+	}
+	if peerForkENR.NextForkVersion != currentForkENR.NextForkVersion {
+		log.WithFields(logrus.Fields{
+			"peerNextForkVersion": peerForkENR.NextForkVersion,
+			"nodeNextForkVersion": currentForkENR.NextForkVersion,
+		}).Debug("Peer matches fork digest but has different next fork version")
+	}
+	return nil
+}
+
+// Adds a fork entry as an ENR record under the eth2EnrKey for
+// the local node. The fork entry is an ssz-encoded enrForkID type
+// which takes into account the current fork version from the current
+// epoch to create a fork digest, the next fork version,
+// and the next fork epoch.
+func addForkEntry(
+	node *enode.LocalNode,
+	genesisTime time.Time,
+	genesisValidatorsRoot []byte,
+) (*enode.LocalNode, error) {
+	currentSlot := helpers.SlotsSince(genesisTime)
+	currentEpoch := helpers.SlotToEpoch(currentSlot)
+
+	// We retrieve a list of scheduled forks by epoch.
+	// We loop through the keys in this map to determine the current
+	// fork version based on the current, time-based epoch number
+	// since the genesis time.
+	currentForkVersion := params.BeaconConfig().GenesisForkVersion
+	scheduledForks := params.BeaconConfig().ForkVersionSchedule
```

Comment:
shouldn't epoch 0 be `params.BeaconConfig().GenesisForkVersion` anyway ? we dont need this line:
```
	currentForkVersion := params.BeaconConfig().GenesisForkVersion
```

Reply:
you mean eliminating that line and just make it part of the scheduledForkVersions map? Genesis isn't a scheduled fork, so just wanted to make that clear

Reply:
Epoch 0 would be the genesis fork version, no ?